Spokestack/SpokestackSpeechRecognizer.swift

```diff
@@ -0,0 +1,199 @@
+            if let error = error {
+                self.context.error = error
+                self.context.dispatch(.error)
+            }
+        }
+        self.task?.receive(completionHandler: self.receive)
+    }
+    
+    private func receive(result: (Result<URLSessionWebSocketTask.Message, Error>)) {
+        self.handle(result, handleResult: { r in
+            if let hypothesis = r.hypotheses.last {
+                self.context.confidence = hypothesis.confidence
+                self.context.transcript = hypothesis.transcript
+                if r.final {
+                    self.context.dispatch(.recognize)
+                }
+            }
+        })
+    }
+
+    private func handle(_ result: Result<URLSessionWebSocketTask.Message, Error>, handleResult: (ASRResult) -> Void) {
+        switch result {
+        case .failure(let error):
+            self.context.error = error
+            self.context.dispatch(.error)
+        case .success(let message):
+            switch message {
+            case .string(let json):
+                do {
+                    guard let jsonData = json.data(using: .utf8) else {
+                        throw SpeechPipelineError.invalidResponse("Could not desearialize the ASR response.")
+                    }
+                    let r = try self.decoder.decode(ASRResult.self, from: jsonData)
+                    if let error = r.error {
+                        throw SpeechPipelineError.failure("Spokestack ASR responded with an error: \(error)")
+                    } else {
+                        handleResult(r)
+                    }
+                } catch let error {
+                    self.context.error = error
+                    self.context.dispatch(.error)
+                }
+            case _:
+                self.configuration.delegateDispatchQueue.async {
+                    self.context.error = SpeechPipelineError.illegalState("Spokestack ASR response with something unknown: \(message)")
+                    self.context.dispatch(.error)
+                }
+            }
+        }
+    }
+}
+
+extension SpokestackSpeechRecognizer: SpeechProcessor {
+    
+    /// Triggered by the speech pipeline, instructing the recognizer to begin streaming and processing audio.
+    public func startStreaming() { }
+
+    /// Triggered by the speech pipeline, instructing the recognizer to stop streaming audio and complete processing.
+    public func stopStreaming() {
+        if self.isActive {
+            self.deactivate()
+        }
+    }
+    
+    /// Receives a frame of audio samples for processing. Interface between the `SpeechProcessor` and `AudioController` components.
+    /// - Parameter frame: Frame of audio samples.
+    public func process(_ frame: Data) {
+        if self.context.isActive {
+            if !self.isActive {
+                self.activate()
+                self.stream(frame)
+            } else if
+                (self.isActive
```

Comment:
The more components that have to track their own activation length, the more I like the approach of activation management being its own stage.

Reply:
an activation management stage is a great idea